x-pack/filebeat/processors/aws_vpcflow/parse_aws_vpc_flow_log.go

```diff
@@ -0,0 +1,177 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package aws_vpcflow
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	jsprocessor "github.com/elastic/beats/v7/libbeat/processors/script/javascript/module/processor"
+)
+
+const (
+	procName = "parse_aws_vpc_flow_log"
+	logName  = "processor." + procName
+)
+
+func init() {
+	processors.RegisterPlugin(procName, New)
+	jsprocessor.RegisterPlugin("ParseAWSVPCFlowLog", New)
+}
+
+var (
+	errValueNotString = errors.New("field must be a string")
+	errInvalidFormat  = errors.New("log did not match the specified format")
+)
+
+type processor struct {
+	config
+	fields          []vpcFlowField
+	log             *logp.Logger
+	expectedIPCount int
+}
+
+// New constructs a new processor built from ucfg config.
+func New(cfg *conf.C) (processors.Processor, error) {
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, fmt.Errorf("fail to unpack the "+procName+" processor configuration: %w", err)
+	}
+
+	return newParseAWSVPCFlowLog(c)
+}
+
+func newParseAWSVPCFlowLog(c config) (*processor, error) {
+	log := logp.NewLogger(logName)
+	if c.ID != "" {
+		log = log.With("instance_id", c.ID)
+	}
+
+	fields, err := parseFormat(c.Format)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse vpc flow log format: %w", err)
+	}
+
+	var ipCount int
+	for _, f := range fields {
+		if f.Type == ipType {
+			ipCount++
+		}
+	}
+
+	return &processor{config: c, fields: fields, expectedIPCount: ipCount, log: log}, nil
+}
+
+func (p *processor) String() string {
+	json, _ := json.Marshal(p.config)
+	return procName + "=" + string(json)
+}
+
+func (p *processor) Run(event *beat.Event) (*beat.Event, error) {
+	err := p.run(event)
+	if err == nil || p.IgnoreFailure || (p.IgnoreMissing && errors.Is(err, mapstr.ErrKeyNotFound)) {
+		return event, nil
+	}
+
+	return event, err
+}
+
+func (p *processor) run(event *beat.Event) error {
+	v, err := event.GetValue(p.Field)
+	if err != nil {
+		return err
+	}
+
+	strValue, ok := v.(string)
+	if !ok {
+		return errValueNotString
+	}
+
+	itr := newWordIterator(strValue)
+	if itr.Count() != len(p.fields) {
```

Comment:
If this is always being done, it seems a little odd to then do a lazy evaluation of the word iterator (viz note on the `wordrIterator` code).

Reply:
Yeah, that does seem like an opportunity for improvement. I'm thinking about removing `wordIterator` for now since it's not providing an huge savings in allocations. And come back in later and see if I can do better.

I was thinking of an alternative implementation of `strings.Fields` that accepts an array instead of returning a slice. It takes advantage of the fact that we know a maximum number of fields to expect such that we can declare an array that's hopefully stack allocated. Like...

`func Fields(word string, indices [maxWords]string) count int`

Reply:
Rather `func Fields(dst []string, s string) int` (or `*[maxWords]string`) to make the change be reflected after the call.